File: frontend/types/job.ts
export type JobType =
  | "Full-time"
  | "Part-time"
  | "Contract"
  | "Internship"
  | "Temporary";

export type CompensationInterval = "yearly" | "monthly" | "weekly" | "hourly";

export type JobLevel =
  | "Entry"
  | "Mid"
  | "Senior"
  | "Lead"
  | "Principal"
  | "Director";

export type JobStatus = "saved" | "applied" | "interview" | "declined";

export interface Job {
  id: number;
  job_url: string | null;
  job_url_direct: string | null;
  title: string;
  company_name: string | null;
  company_id: number | null;
  description: string | null;
  company_url: string | null;
  company_url_direct: string | null;
  location_city: string | null;
  location_state: string | null;
  location_country: string | null;
  compensation_min: number | null;
  compensation_max: number | null;
  compensation_currency: string | null;
  compensation_interval: CompensationInterval | null;
  job_type: JobType[] | null;
  date_posted: Date | null;
  is_remote: boolean | null;
  listing_type: string | null;
  job_level: JobLevel | null;
  job_function: string | null;
  company_industry: string | null;
  company_headquarters: string | null;
  company_employees_count: string | null;
  applicants_count: number | null;
  // Company employee size fields (from Company model)
  company_size_min: number | null;
  company_size_max: number | null;
  company_size_on_linkedin: number | null;
  emails: string[] | null;
  
  // LLM-parsed fields from job description
  required_skills: string[] | null;
  preferred_skills: string[] | null;
  required_years_experience: number | null;
  required_education: string | null;
  preferred_education: string | null;
  responsibilities: string[] | null;
  benefits: string[] | null;
  work_arrangement: string | null;
  team_size: string | null;
  technologies: string[] | null;
  culture_keywords: string[] | null;
  summary: string | null;
  job_categories: string[] | null;
  independent_contractor_friendly: boolean | null;
  parsed_salary_currency: string | null;
  parsed_salary_min: number | null;
  parsed_salary_max: number | null;
  compensation_basis: string | null;
  location_restrictions: string[] | null;
  exclusive_location_requirement: boolean | null;
  
  // DSPy-parsed fields from job description
  is_python_main: boolean | null;
  contract_feasible: boolean | null;
  relocate_required: boolean | null;
  specific_locations: string[] | null;
  accepts_non_us: boolean | null;
  screening_required: boolean | null;
  
  created_at: Date;
  updated_at: Date;
}

export interface Company {
  id: number;
  linkedin_url: string;
  linkedin_internal_id: string | null;
  name: string;
  description: string | null;
  has_own_products: boolean | null;
  is_recruiting_company: boolean | null;
  website: string | null;
  industry: string | null;
  company_size_min: number | null;
  company_size_max: number | null;
  company_size_on_linkedin: number | null;
  hq_country: string | null;
  hq_city: string | null;
  hq_state: string | null;
  hq_postal_code: string | null;
  company_type: string | null;
  founded_year: number | null;
  tagline: string | null;
  universal_name_id: string | null;
  profile_pic_url: string | null;
  background_cover_image_url: string | null;
  specialities: string[] | null;
  locations: Array<Record<string, unknown>> | null;
  created_at: Date;
  updated_at: Date;
}

export interface JobDetail extends Job {
  company: Company | null;
}

export interface SavedJob {
  id: number;
  user_id: number;
  job_id: number;
  status: JobStatus;
  notes: string | null;
  created_at: Date;
  updated_at: Date;
  job: JobDetail;
}

export interface SavedJobListResponse {
  saved_jobs: SavedJob[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface SavedJobCheckResponse {
  is_saved: boolean;
  saved_job_id: number | null;
  status: JobStatus | null;
}

File: backend/app/schemas/job.py
from __future__ import annotations

from datetime import date, datetime
from math import ceil
from typing import Sequence

from pydantic import BaseModel, ConfigDict, field_serializer

from app.schemas.company import CompanyResponse


class JobBaseResponse(BaseModel):
    id: int
    job_url: str | None = None
    job_url_direct: str | None = None
    title: str
    company_name: str | None = None
    company_id: int | None = None
    company_url: str | None = None
    company_url_direct: str | None = None
    location_city: str | None = None
    location_state: str | None = None
    location_country: str | None = None
    compensation_min: float | None = None
    compensation_max: float | None = None
    compensation_currency: str | None = None
    compensation_interval: str | None = None
    job_type: list[str] | None = None
    date_posted: date | None = None
    is_remote: bool | None = None
    listing_type: str | None = None
    job_level: str | None = None
    job_function: str | None = None
    company_industry: str | None = None
    company_headquarters: str | None = None
    company_employees_count: str | None = None
    applicants_count: int | None = None
    # Company employee size fields (from Company model)
    company_size_min: int | None = None
    company_size_max: int | None = None
    company_size_on_linkedin: int | None = None
    emails: list[str] | None = None
    
    # LLM-parsed fields from job description
    required_skills: list[str] | None = None
    preferred_skills: list[str] | None = None
    required_years_experience: int | None = None
    required_education: str | None = None
    preferred_education: str | None = None
    responsibilities: list[str] | None = None
    benefits: list[str] | None = None
    work_arrangement: str | None = None
    team_size: str | None = None
    technologies: list[str] | None = None
    culture_keywords: list[str] | None = None
    summary: str | None = None
    job_categories: list[str] | None = None
    independent_contractor_friendly: bool | None = None
    parsed_salary_currency: str | None = None
    parsed_salary_min: float | None = None
    parsed_salary_max: float | None = None
    compensation_basis: str | None = None
    location_restrictions: list[str] | None = None
    exclusive_location_requirement: bool | None = None
    
    # DSPy-parsed fields from job description
    is_python_main: bool | None = None
    contract_feasible: bool | None = None
    relocate_required: bool | None = None
    specific_locations: list[str] | None = None
    accepts_non_us: bool | None = None
    screening_required: bool | None = None
    
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @field_serializer("created_at", "updated_at", when_used="json")
    def serialize_datetime(self, value: datetime) -> str:
        return value.isoformat()

    @field_serializer("date_posted", when_used="json")
    def serialize_date(self, value: date | None) -> str | None:
        return value.isoformat() if value else None


class JobListItemResponse(JobBaseResponse):
    pass


class JobResponse(JobBaseResponse):
    description: str | None = None


class JobDetailResponse(JobResponse):
    company: CompanyResponse | None = None


class JobListResponse(BaseModel):
    jobs: list[JobListItemResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

    model_config = ConfigDict(from_attributes=True)

    @classmethod
    def from_results(
        cls,
        jobs: Sequence[JobListItemResponse],
        total: int,
        page: int,
        page_size: int,
    ) -> "JobListResponse":
        total_pages = ceil(total / page_size) if total else 0
        return cls(
            jobs=list(jobs),
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
        )
